src/rgw/rgw_rados.cc

```diff
@@ -11996,3 +12030,20 @@ librados::Rados* RGWRados::get_rados_handle()
   }
 }
 
+int rgw_compression_info_from_attrset(map<string, bufferlist>& attrs, bool& need_decompress, RGWCompressionInfo& cs_info) {
+  if (attrs.find(RGW_ATTR_COMPRESSION) != attrs.end()) {
+    bufferlist::iterator bliter = attrs[RGW_ATTR_COMPRESSION].begin();
+    try {
+      ::decode(cs_info, bliter);
+    } catch (buffer::error& err) {
+      return -EIO;
+    }
+    need_decompress = true;
```

Comment:
To avoid potential issues I'd suggest to check retrieved compression type for none too.